pkg/engine/imageVerify.go

```diff
@@ -220,54 +219,77 @@ func (iv *imageVerifier) markImageVerified(imageVerify v1.ImageVerification, rul
 
 	if imageVerify.Required {
 		isImageVerified := ruleResp.Status == response.RuleStatusPass
-		patch, err := makeImageVerifiedPatch(imageInfo, digest, isImageVerified)
-		if err == nil {
-			ruleResp.Patches = [][]byte{patch}
-		} else {
+		hasAnnotations := len(iv.policyContext.NewResource.GetAnnotations()) > 0
+		patches, err := makeImageVerifiedPatches(imageInfo, digest, isImageVerified, hasAnnotations)
+		if err != nil {
 			iv.logger.Error(err, "failed to create patch", "image", imageInfo.String())
+		} else {
+			ruleResp.Patches = append(ruleResp.Patches, patches...)
 		}
 	}
 
 	return ruleResp
 }
 
 func hasImageVerifiedAnnotationChanged(ctx *PolicyContext, name, digest string) bool {
-	if reflect.DeepEqual(ctx.OldResource, &unstructured.Unstructured{}) ||
-		reflect.DeepEqual(ctx.NewResource, &unstructured.Unstructured{}) {
+	if reflect.DeepEqual(ctx.NewResource, &unstructured.Unstructured{}) ||
+		reflect.DeepEqual(ctx.OldResource, &unstructured.Unstructured{}) {
 		return false
 	}
 
-	key := makeAnnotationKey(name, digest)
+	key := makeAnnotationKey(name)
 	newValue := ctx.NewResource.GetAnnotations()[key]
 	oldValue := ctx.OldResource.GetAnnotations()[key]
 	return newValue != oldValue
 }
 
-func makeImageVerifiedPatch(imageInfo kubeutils.ImageInfo, digest string, verified bool) ([]byte, error) {
-	var patch = make(map[string]interface{})
-	annotationKey := makeAnnotationKeyForJSONPatch(imageInfo.Name, digest)
-	patch["op"] = "add"
-	patch["path"] = annotationKey
-	patch["value"] = strconv.FormatBool(verified)
-	return json.Marshal(patch)
-}
+func makeImageVerifiedPatches(imageInfo kubeutils.ImageInfo, digest string, verified, hasAnnotations bool) ([][]byte, error) {
+	var patches [][]byte
+	if !hasAnnotations {
+		var addAnnotationsPatch = make(map[string]interface{})
+		addAnnotationsPatch["op"] = "add"
+		addAnnotationsPatch["path"] = "/metadata/annotations"
+		addAnnotationsPatch["value"] = map[string]string{}
+		patchBytes, err := json.Marshal(addAnnotationsPatch)
+		if err != nil {
+			return nil, err
+		}
 
-func makeAnnotationKeyForJMESPath(imageName, imageDigest string) string {
-	key := makeAnnotationKey(imageName, imageDigest)
-	return "request.object.metadata.annotations." + `"` + key + `"`
+		patches = append(patches, patchBytes)
+	}
+
+	imageData := &ImageVerificationMetadata{
+		Verified: verified,
+		Digest:   digest,
+	}
+
+	data, err := json.Marshal(imageData)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to marshal metadata value: %v", imageData)
+	}
+
+	var addKeyPatch = make(map[string]interface{})
+	annotationKey := makeAnnotationKeyForJSONPatch(imageInfo.Name)
```

Comment:
@JimBugwadia what would happen if there are two images with the same name but different registries/paths/digests?

Reply:
@samj1912 - good point. Logged https://github.com/kyverno/kyverno/issues/3754.